### PR TITLE
Adding name convention validator for QHub project name

### LINKED ISSUE #626: [bug] `qhub-deploy` fails for azure project with dashes in the name
## Describe the bug

`qhub-deploy` fails to deploy to Azure when using a project with dashes in the name. The error message is

```
[terraform]: Error: Failed to get existing workspaces: Error retrieving keys for Storage Account "qhub-test-tomdevtz9o1ryx": storage.AccountsClient#ListKeys: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceNotFound" Message="The Resource 'Microsoft.Storage/storageAccounts/qhub-test-tomdevtz9o1ryx' under resource group 'qhub-test-tom-dev' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix"
```

A few pieces of information:

project name: `qhub-test-tom`. 
stroage account postfix: `tz9o1ryx`.
storage account name:  `qhubtesttomdevtz9o1ryx`

But in `state.tf`, we have

```
terraform {
  backend "azurerm" {
    resource_group_name = "qhub-test-tom-dev"
    # storage account must be globally unique
    storage_account_name = "qhub-test-tomdevtz9o1ryx"
    container_name       = "qhub-test-tom-devstate"
    key                  = "terraform/qhub-test-tom-dev.tfstate"
  }
}
```

The storage account has dashes in it (which I assume aren't allowed by Azure).

This created an Azure Storage Account name 

## How can we help?
Help us help you.
* What are you trying to achieve?
* How can we reproduce the problem?
* What is the expected behaviour?
* And what is currently happening?
* Any error messages?
> If helpful, please add any screenshots and relevant links to the description.

## Your environment
Describe the environment in which you are experiencing the bug.
> Include your conda version (use `conda --version`), k8s and any other relevant details.

```
$ conda list qhub
# packages in environment at /home/taugspurger/miniconda3/envs/qhub:
#
# Name                    Version                   Build  Channel
qhub                      0.3.11             pyhd8ed1ab_0    conda-forge
```

### PR DESCRIPTION
**Pull request by viniciusdc**
In short, as QHub utilizes different cloud providers each one has a number of constraints that must be considered while naming resources. Because of that, QHub currently uses some naming conventions to better address these outstanding restrictions. Using this validator we will avoid issues with QHub deployment as well give more feedback to the end user.

QHub [naming convetion](https://docs.qhub.dev/en/latest/source/installation/usage.html?highlight=naming#project-naming-convention) used as base.

close #626

### CODE REVIEW

File: qhub/schema.py
Comment (tylerpotts) [inline qhub/schema.py:371]:
```suggestion
        to be followed. First, ensure your name is compatible with the specific one for
```

File: qhub/schema.py
Comment (tylerpotts) [inline qhub/schema.py:377]:
```suggestion
        - If using AWS: names should not start with the string "aws";
```

File: qhub/schema.py
Comment (tylerpotts) [inline qhub/schema.py:375]:
```suggestion
        - Special characters are not allowed, such as "!@#$%^&*()";
```

File: qhub/schema.py
Comment (tylerpotts) [inline qhub/schema.py:372]:
```suggestion
        your chosen Cloud provider. In addition, the QHub project name should also obey the following
```

File: qhub/schema.py
Comment (tylerpotts) [inline qhub/schema.py:383]:
```suggestion
        elif re.findall(r"^(?!aws)[A-Za-z0-9][^\|\\?_]*[A-Za-z0-9]$", value):
```
Dashes are acceptable in names. The previous regular expression looks for dashes and erroniously was throwing errors, causing all the tests to crash

**Review by tylerpotts — CHANGES_REQUESTED** [review]
Edited text for grammar/tone.

Using ALL CAPS FEELS LIKE YELLING :)

Also, the phrase "for starters" has connotations that there's a big list of things incoming, and sends the wrong message when indicating the user did something wrong (this could be a person preference thing)

Also suggested edits to allow `-` in the names

File: qhub/schema.py
Comment (viniciusdc) [reply qhub/schema.py:383]: Thank you Tyler for the review, I red the special characters not allowed in the convention and thought that `-` was a problem to. Thank you, I will commit those changes.

File: qhub/schema.py
Comment (tylerpotts) [inline qhub/schema.py:377]: Where does this come from? It looks like the current aws test has a project name `aws-test`

Comment (viniciusdc) [reply qhub/schema.py:377]: It is the description, but I don't remember why we need to avoid that. I remember a discussion about that a really long time ago... Anyways I will need to add another step in that validation because the [Azure storage name](https://github.com/toddkitta/azure-content/blob/master/articles/guidance/guidance-naming-conventions.md#naming-rules-and-restrictions) must not contain any dashes

File: qhub/schema.py
Comment (viniciusdc) [inline qhub/schema.py:383]: @tarundmsharma so right now this regex is not considering all dashes in the project name, which is not something we would like to have for all providers... only for Azure. So we need to add a conditional that checks the target provider and use the correct regex based on it.

**Comment by costrouc** [comment]
@tarundmsharma this PR is failing due to the tests having `aws` in the project name in the beginning. For example https://github.com/Quansight/qhub/blob/main/tests/test_render.py#L11. I'd suggest renaming all the project names in tests to `pytest-<provider>` instead of `<pytest>-provider` this way the tests will pass.

File: qhub/schema.py
Comment (costrouc) [reply qhub/schema.py:384]: I'd rather have a simple regex and rules for users on what can be in a project name so that we don't have to have so many exceptions. For example:
  - must begin with a letter `[a-z][A-Z]`
  - cannot have name begin with `aws`
  - name can only contain `[a-zA-Z0-9]`

Can you make this the rule for checking?  Is this sane?

**Comment by viniciusdc** [comment]
This is now blocked by https://github.com/Quansight/qhub/pull/837, we need to change our test project names in order to accept this incoming changes (at least to get a passing test) @costrouc

**Comment by iameskild** [comment]
I will fix as part of my work for #874.

**Comment by iameskild** [comment]
@viniciusdc once #927 is approved and merged, I can go back and resolve any merge conflicts here 👍

**Comment by viniciusdc** [comment]
This might need to be considered as well https://github.com/Quansight/qhub/blob/90e344810834038e10b22906ebb3222f29a405e2/qhub/template/hooks/pre_gen_project.py

**Comment by iameskild** [comment]
@viniciusdc this is ready for another review when you get the chance, thanks :)

**Comment by viniciusdc** [comment]
This is now up-to-date with main, though this now requires a review of our naming conventions to continue. I will open a tracking issue to research this and update here once finished.

**Comment by costrouc** [comment]
@viniciusdc could you update the tests?

**Pull request merged**
